### ISSUE TITLE
In some scenarios, opening Atom on Linux results in 100% CPU usage of a core

### ISSUE THREAD

**Issue by joefitzgerald** [atom/atom#13317]
* When some Linux users open Atom via .desktop files or via the gnome launcher, they experience 100% CPU usage due to the way we are invoking the shell to patch the user's environment
* This does not happen for all Linux users
* This does seem to happen consistently for users on Arch Linux
* The shell is invoked with `detached: false` and using the `-i` flag for an interactive shell
  * The intent of shelling out to get the environment is to run the same shell initialization scripts that are run when a user opens a terminal window and invokes atom from there
  * Removing the `-i` flag is not the solution, because it doesn't result in the desired behavior; it does result in the shell exiting as expected without consuming 100% of a CPU core
  * Adding `detached: true` to the options used for `child_process.spawn` _does_ fix the issue and retains the desired behavior, but it requires additional code to manage the process lifecycle if Atom exits or an Atom window is reloaded while the shell process is still running

The below is a copy of https://github.com/atom/atom/issues/13084#issuecomment-262507036, included here to open up a new issue:

@phw says:

~~~
I debugged this a bit and can reproduce it with the following test script (extracted from [src/update-process-env.js](https://github.com/atom/atom/blob/1b2d2f029945c58cf4cd474ef1423856b35cf26c/src/update-process-env.js)):

```javascript
const childProcess = require('child_process')
const cmd = 'command env'
const shell = process.env.SHELL  // zsh or bash doesn't matter, behavior is the same
// const shell = '/bin/bash'

const child = childProcess.spawn(shell, ['-ilc', cmd], {
   encoding: 'utf8',
  //  detached: true,
  stdio: ['ignore', 'pipe', process.stderr]
})
child.on('error', (e) => {
  console.log('error', e)
})
child.stdout.on('data', (data) => {
  console.log(data.toString('utf8'))
})
child.on('close', (code, signal) => {
  console.log('close', code, signal)
})
```

Launching this directly from a shell works, even if I set an empty environment like so:

    env -i SHELL=/bin/bash node spawn.js

Launching the same from a desktop file in Gnome 3.22 leads to the shell task endlessly consuming 100% of CPU. I used the following desktop file:

```
[Desktop Entry]
Name=Spawn Node Test
Exec=env SHELL=/bin/bash node /home/phw/devel/spawn.js
Icon=atom
Type=Application
```

Two things help to solve this issue:

a) Enable `detached: true` when spawning the child process
b) Remove the `-i` flag from the shell call

I think b) would actually be correct anyway, since this is in no way an interactive shell.

EDIT: Just to be clear: I still don't know exactly *why*  this is happening, I just know the issue can be circumvented removing the `-i` flag.

**Comment by phw** [comment]
Maybe some other interesting fact: If Atom is launched via the desktop file and consequently spawns this non-terminating shell process, some actions (New file, About, maybe others) do not react. That is you trigger the action (via menu, keyboard shortcut, command palette, doesn't matter) but there is no action.

However, once you kill the spawned shell process the triggered actions get executed. E.g. you hit two times `Ctrl+N` and also start "Application: About" via command palette, nothing visible happens. You kill the spawned shell process and suddenly two new file editors and the about window are opened.

Maybe somebody has an idea how this is related.

**Comment by garretraziel** [comment]
It is happening to me on my system, when I click on "restart" after package update (it was update of `nucleus-dark-ui`). Fedora 25, Atom 1.12.5 installed from rpm downloaded directly from atom.io.

**Comment by joefitzgerald** [comment]
@phw please try building from source (use the master branch) and let me know if this has resolved your issue.

**Comment by joefitzgerald** [comment]
@phw or others affected by this: in order to release a hotfix, we want one of you to build from source on Linux and verify the issue is fixed. Could you please assist us in doing this?

**Comment by phw** [comment]
I'd love to test this, but I likely won't be able to do so before Monday.

**Comment by JanEerik** [comment]
Just built from source and seems to have fixed the problem in my case (running ubuntu 16.04 LTS).

**Comment by ezickler** [comment]
Would like to try but i'm unable to compile atom. I think some nodejs problem with fedora but tried with packagemanger and nodejs installed over nvm. Dosn't work for me. Sry. I don't get this node stuff.

**Comment by PAStheLoD** [comment]
Hello,

Just to make sure this is the same bug, I get the same behavior, with the same `bash -ilc command env` trapped in an infinite signal handling loop ([see also](http://stackoverflow.com/questions/13718394/what-should-interactive-shells-do-in-orphaned-process-groups)), with bash 4.4:

```
strace: Process 19539 attached
--- SIGTTIN {si_signo=SIGTTIN, si_code=SI_USER, si_pid=19539, si_uid=1000} ---
rt_sigaction(SIGTTIN, {SIG_IGN, [], SA_RESTORER, 0x7fba456ab860}, {SIG_DFL, [], SA_RESTORER, 0x7fba456ab860}, 8) = 0
ioctl(255, TIOCGPGRP, [19683])          = 0
rt_sigaction(SIGTTIN, {SIG_DFL, [], SA_RESTORER, 0x7fba456ab860}, {SIG_IGN, [], SA_RESTORER, 0x7fba456ab860}, 8) = 0
kill(0, SIGTTIN)                        = 0
...
```

If I try to strace the start of atom `strace -f -e execve -s 300 -o /tmp/atom.trace atom .` in my `$HOME`, atom starts normally (maybe because using strace changes the process group structure).

**Comment by phw** [comment]
@joefitzgerald I finally was able to test this, looking very good 👍 I tried with latest master, starting Atom from the .desktop file works as expected. Just to be sure I also installed the revision just before the merge of your fix (03cc5033b7d4030f23c6a393dd32aa0c882f306b), and as expected the issue was there again.

Thanks for this fix :)
